### PR TITLE
Prevent goroutine leak in http service for persistent query connections

### PR DESCRIPTION
**Pull request by aaronknister**
###### Required for all non-trivial PRs
- [ ] Rebased/mergable
- [ ] Tests pass
- [ ] CHANGELOG.md updated
- [ ] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

###### Required only if applicable
_You can erase any checkboxes below this note if they are not applicable to your Pull Request._
- [ ] [InfluxQL Spec](https://github.com/influxdata/influxdb/blob/master/influxql/README.md) updated

### CODE REVIEW

File: services/httpd/handler.go [hunk 1]
```diff
@@ -292,14 +292,19 @@ func (h *Handler) serveQuery(w http.ResponseWriter, r *http.Request, user *meta.
...
 		notify := notifier.CloseNotify()
 		go func() {
-			<-notify
-			close(closing)
+			// Wait for either the request to close "closing"
+			// on exit, or for the http close notifier to kick in
+			select {
+			case <-closing:
+			case <-notify:
+				close(closing)
```
Comment (jsternberg) [inline services/httpd/handler.go]: This will cause a panic due to a double close.

Comment (aaronknister) [inline services/httpd/handler.go]: 10-4. I just realized that looking at the circleci tests. I've tried it a different way that should avoid the double close.

**Comment by jsternberg** [comment]
Most likely, what you're trying to do isn't possible. Pipelining and `CloseNotify()` are documented as not compatible with each other. Can you explain to me how you're trying to use the server where you need to use HTTP pipelining? Can you show me a reproducer that causes the leaked goroutine?

It might be best just to have the server send the `Connection: close` header to force a new connection and prevent this issue.

**Comment by aaronknister** [comment]
I can accept that. The latest patch seems to work, so far. The reproducer is Grafana running a query with a relatively high refresh interval (e.g. 10s).

**Comment by jsternberg** [comment]
I think unless there's a required performance reason it's better to just force the connection to close. The use of `CloseNotify()` like that is undefined and seems to differ based on the Go version used, so I'm cautious to include this code. I'm going to review it anyway though in case we want to include it as a defensive coding mechanism.

Does Grafana use pipelined queries?

File: services/httpd/handler.go [hunk 2]
```diff
@@ -292,11 +292,18 @@ func (h *Handler) serveQuery(w http.ResponseWriter, r *http.Request, user *meta.
 
 	// Make sure if the client disconnects we signal the query to abort
 	closing := make(chan struct{})
+	reqDone := make(chan struct{})
```
Comment (jsternberg) [inline services/httpd/handler.go]: Can you rename this to `done`?

Comment (jsternberg) [inline services/httpd/handler.go]: Also, can you only do this when a `CloseNotifier` is used? I think `CloseNotify` is always used in practicality, but I would like that section of the code moved into the `if` statement anyway with a comment about why it is needed.

**Comment by aaronknister** [comment]
I don't think Grafana is pipelining queries based on what I've seen, but it is reusing a given TCP socket for multiple serial queries.

File: services/httpd/handler.go (hunk 2 above)
Comment (aaronknister) [inline services/httpd/handler.go]: Absolutely, I'll do that.

**Comment by jsternberg** [comment]
@aaronknister is it using a single connection for multiple queries (by separating statements with a semi-colon) or using the same underlying socket to make multiple HTTP calls? Reusing a TCP socket would be the later.

Either way, I think this change should work fine. I was thinking over the weekend about adding this anyway for safety. Can you update the changelog and sign the CLA? Be sure to give yourself credit in the changelog.

Thanks.

**Comment by jsternberg** [comment]
Something I didn't notice. Can you make this change on master? If we think this is worthy of adding to a bugfix release, we'll cherry pick it to the 0.12 branch.

**Comment by aaronknister** [comment]
I've opened #6413 with the change on master. Should I close out this PR?

**Pull request closed without merging**